{% extends "home/base.html" %}
{% load staticfiles %}
{% block addon_css %}
<link rel="stylesheet" href="{% static 'home/css/jquery.dataTables.min.css' %}" type="text/css" />
<link rel="stylesheet" href="{% static 'home/css/nv.d3.min.css' %}" type="text/css" />
<!-- <link rel="stylesheet" href="{% static 'home/css/structure_browser.css' %}" type="text/css" /> -->
<!-- <link href="{% static 'home/css/residue_table.css' %}" rel="stylesheet"> -->
<!-- <link href="{% static 'home/css/alignment.css' %}" rel="stylesheet"> -->
<style>
table.dataTable thead tr th {
    word-wrap: break-word;
    word-break: break-all;
}
table.dataTable tbody tr td {
    word-wrap: break-word;
    word-break: break-all;
}
</style>
{% endblock %}
{% block content %}

<h3>Estimated economic burden on the NHS of GPCR targeting drugs</h3><br>
<div id="burden" class="chart_container" style="display: none;">
       <div id="burden"><svg id="burden" style="width:1050;height:500px;"></svg></div>

<div class="row">
<div class="col-sm-3"></div>
  <div class="col-md-6">
     <button id="update">add scaling factor for drugs targeting polymorphic putative functional sites</button>
     <input id="updater" type="text" maxlength="5" size="6" value="1.0" name="cutoff"
  </div>
  <div class="col-sm-3"></div>
</div>
<br><br><br>
<h4>Calculations</h4>
<p>The economic burden was estimated by:</p>
<p style="text-align: center;"><em>estimated economic burden per drug (&pound;) <strong>=</strong> NHS cost per drug per year (&pound;) <strong>x</strong></em></p>
<p style="text-align: center;"><em>% individuals with a  MV in a functional site of the respective drug targets</em></p>
<!-- <p style="text-align: center;"><em>correction or scaling factor&nbsp;</em></p> -->
<p>where:</p>
<ul style="list-style-type: circle;">
<li>The average NHS cost is the average yearly cost over a 4-year period (2013-2016) per GPCR drug that is listed (n=279)
</li>
<li>% Individuals is the percentage of affected individuals with a missense variant in a functional site of the respective drug target(s) (2,504 individuals from 1,000 Genomes Project phase 3 genotype data)</li>
<li>The fraction of affected individuals was calculated by considering (a) individuals who have a variation in a functional site in both alleles (homozygous) and (b) individuals if they have a variation in a functional site in at least one allele (i.e. homozygous and heterozygous).
</li>
<li>Functional sites are distinguished into (a) known functional sites that included ligand binding, effector binding, PTM site, sodium binding site and microswitches and (b) putative functional site as those predicted to be deleterious based on SIFT or PolyPhen.</li>
<!-- <li>the correction or scaling factor combines the contributing effects including other targets, metabolism ADR, dose/efficacy effects, buffering mutations, depletion of susceptible effects (decreased risk of adverse events over time, associated care costs such as for mental health illness) <em>etc</em>.</li> -->
</ul>
<p>For these estimates, we assume that for the NHS data each prescription is a unique individual.</p>
<br><br>
<h3>Sales and variation information per drug</h3><br>
    <!-- Download button -->
    <a class="btn btn-sm btn-primary" href="download" role="button">
    <span class="glyphicon glyphicon-download"></span> Download</a>
<div class="row">
    <div class="col-md-11 text-center">
        <div style="padding-top: 10px; font-size: 11px;white-space: nowrap;">
            <table width="100%" id="drugdata">
                <thead>
                    <tr>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="amino acid substitution, wildtype to mutant" data-placement="top">Drug Name</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="Category of mutation" data-placement="top">GPCR Drug Targets</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="ExAC allele frequencies, as calculated as allele count/allele number" data-placement="top">Sales Average</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="number of alleles with this variant" data-placement="top">Prescription Average</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="total number of alleles that were able to be called" data-placement="top">Item Cost</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="deleterious: <= 0.05<br>tolerated: > 0.05" data-placement="top">Putative Functional Site Variants</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="deleterious: <= 0.05<br>tolerated: > 0.05" data-placement="top">Known Functional Site Variants</th>
                        <th style="text-align:center" class="citation-tooltip" target="_blank" data-toggle="tooltip"  data-container="body" data-html="true" data-original-title="deleterious: <= 0.05<br>tolerated: > 0.05" data-placement="top">BNF section</th>
                      </tr>
                      <tr>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                          <th style="text-align:center"></th>
                      </tr>
                      <tr>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                        <th style="text-align:center"></th>
                      </tr>
                </thead>
                <tfoot>
                    <tr>
                      <th style="text-align:center">Drug Name</th>
                      <th style="text-align:center">GPCR Drug Targets</th>
                      <th style="text-align:center">Sales Average</th>
                      <th style="text-align:center">Prescription Average</th>
                      <th style="text-align:center">Item Cost</th>
                      <th style="text-align:center">Known Functional Site Variants</th>
                      <th style="text-align:center">Putative Functional Site Variants</th>
                      <th style="text-align:center">BNF section</th>
                    </tr>
                </tfoot>
                {% for drug in drug_data %}
                <!-- add: .select_related.all -->
                <tr>
                  <td style="text-align:center"><a href='/drugs/nhs/{{drug.drugname|safe}} '> {{drug.drugname|safe}}</a></td>
                  <td>{% for target in drug.targets %}
                  <a href='/mutational_landscape/protein/{{ target|add:"_human" }} '>{{ target }}</a>
                  {% endfor %}</td>
                  <td>{{drug.average_cost}}</td>
                  <td>{{drug.average_items}}</td>
                  <td>{{drug.item_cost}}</td>
                  <td>{{drug.putative_func}}</td>
                  <td>{{drug.known_func}}</td>
                  <td>{% for section in drug.section %}
                  <a href='/drugs/nhs/section/{{ section }} '>{{ section }}</a>
                  {% endfor %}</td>
                </tr>
                {% endfor %}
            </table>
        </div>
      </div>
    </div>
<br><br>
{% endblock %}
{% block addon_js %}
<script src="{% static 'home/js/saveSvgAsPng.js' %}"></script>
<script src="{% static 'home/js/d3.min.js' %}"></script>
<script src="{% static 'home/js/nv.d3.min.js' %}"></script>
<script src="{% static 'home/js/jquery.dataTables.min.js' %}"> </script>
<script src="{% static 'home/js/dataTables.tableTools.min.js' %}"> </script>
<script src="{% static 'home/js/jquery.dataTables.columnFilter.js' %}"> </script>

<script type="text/javascript" charset="utf-8">
        $(document).ready(function() {
            var table = $('#drugdata').DataTable({
                // 'scrollX': true,
                /*// 'scrollY': $(window).height()-300,*/
                // 'scrollCollapse': true,
                'paging': true,
                'lengthMenu': [[20, 50, 100, -1], [20, 50, 100, "All"]],
                'orderCellsTop': true,
                "bInfo" : false,
                'autoWidth': false,
                // 'sScrollX': "100%",
                // 'sScrollXInner': "100%",
                'order': [[3, "desc"]],
                'dom': 'T<"clear">lfrtip',
                'columnDefs': [
                  {"width": "10%", targets: 0},
                  {"width": "20%", targets: 1},
                  {"width": "10%", targets: 2},
                  {"width": "10%", targets: 3},
                  {"width": "30%", targets: 4},
                  {"width": "10%", targets: 5},
                  {"width": "10%", targets: 6},
                  {"width": "5%", targets: 7}],
                'tableTools': {
                    "sRowSelect": "multi",
                    "aButtons": []
                },
                initComplete: function () {
                    $('#drugdata').dataTable().columnFilter({
                        'sPlaceHolder': "head:after",
                        'aoColumns': [
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                            { type: "text" },
                        ]
                    });
                }

            });
        });
    </script>
<script>
// https://github.com/novus/nvd3/tree/master/examples
var chart ;
var chartData;

window.onload = function () {

  nv.addGraph(function() {
      var datum = {{data|safe}}

      chart = nv.models.multiBarChart()
        .reduceXTicks(false)   //If 'false', every single x-axis tick label will be rendered.
        // .rotateLabels(0)      //Angle to rotate x-axis labels.
        .showControls(true)   //Allow user to switch between 'Grouped' and 'Stacked' mode.
        // .groupSpacing(0.1)    //Distance between each group of bars.
        .stacked(true)
        .margin({"left": 120, "right": 50, "top": 100, "bottom": 70,})
      ;

      chart.xAxis
          // .axisLabel("correction or scaling factor")
          // .showMaxMin(true)
          // .tickFormat(d3.format(',.1f'))

      chart.yAxis
          .axisLabel("estimated economic burden per year (GBP, £)")
          .axisLabelDistance(35)
          .tickFormat(d3.format(','));

      chart.multibar.dispatch.on("elementClick", function(e) {
        link = e['data']['key']
          window.open ('/drugs/nhs/section/'+link,'_self',false)
      })

      chartData = d3.select('#burden svg').datum(datum)
      chartData.transition().duration(500).call(chart);

      nv.utils.windowResize(chart.update);

      return chart;
  });

  var raw = {{data|safe}};
  function update(cf) {
    var cf = document.getElementById('updater').value

    var rawLength = raw.length;
    for (var i = 0; i < rawLength; i++) {

      var kk = raw[i]['values'].length;
      raw[i]['values'][kk] = {'y': raw[i]['values'][3]['y']*cf, 'x': cf}
    }

    // Update the SVG with the new data and call chart
    chartData.datum(raw).transition().duration(5).call(chart);
    nv.utils.windowResize(chart.update);
};

d3.select("#update").on("click", update);

    /* Inspired by Lee Byron's test data generator. */
    function stream_layers(n, m, o) {
      if (arguments.length < 3) o = 0;
      function bump(a) {
        var x = 1 / (.1 + Math.random()),
            y = 2 * Math.random() - .5,
            z = 10 / (.1 + Math.random());
        for (var i = 0; i < m; i++) {
          var w = (i / m - y) * z;
          a[i] += x * Math.exp(-w * w);
        }
      }
      return d3.range(n).map(function() {
          var a = [], i;
          for (i = 0; i < m; i++) a[i] = o + o * Math.random();
          for (i = 0; i < 5; i++) bump(a);
          return a.map(stream_index);
        });
    }

    /* Another layer generator using gamma distributions. */
    function stream_waves(n, m) {
      return d3.range(n).map(function(i) {
        return d3.range(m).map(function(j) {
            var x = 20 * j / m - i / 3;
            return 2 * x * Math.exp(-.5 * x);
          }).map(stream_index);
        });
    }

    function stream_index(d, i) {
      return {x: i, y: Math.max(0, d)};
    }
  //Generate some nice data.
  function exampleData(sf) {
    return stream_layers(10,sf,0.1).map(function(data, i) {
      return {
        key: 'Stream #' + i,
        values: data
      };
    });
  }

  $(document).ready(function () {
      $('#burden').css("fill", '#000000');
      $('#burden').css("display", '');
      // draw('1');

  });
};
</script

{% endblock %}
